service/storage_service.cc

```diff
@@ -4412,7 +4412,62 @@ future<> storage_service::move(token new_token) {
 
 future<std::vector<storage_service::token_range_endpoints>>
 storage_service::describe_ring(const sstring& keyspace, bool include_only_local_dc) const {
-    return locator::describe_ring(_db.local(), _gossiper, keyspace, include_only_local_dc);
+    size_t nr_tablet_tables = 0;
+    co_await _db.local().get_tables_metadata().for_each_table_gently(coroutine::lambda([&] (table_id id, lw_shared_ptr<replica::table> table) -> future<> {
+        auto s = table->schema();
+        if (s->ks_name() == keyspace && table->uses_tablets()) {
+            ++nr_tablet_tables;
+        }
+        co_return;
+    }));
+    if (nr_tablet_tables) {
+        throw std::runtime_error(format("The keyspace {} has tablet table. Query describe_ring with the table parameter!", keyspace));
+    }
+    co_return co_await locator::describe_ring(_db.local(), _gossiper, keyspace, include_only_local_dc);
+}
+
+future<std::vector<dht::token_range_endpoints>>
+storage_service::describe_ring_for_table(const sstring& keyspace_name, const sstring& table_name) const {
+    slogger.debug("describe_ring for table {}.{}", keyspace_name, table_name);
+    auto& t = _db.local().find_column_family(keyspace_name, table_name);
+    if (!t.uses_tablets()) {
+        auto ranges = co_await describe_ring(keyspace_name);
+        co_return ranges;
+    }
+    table_id tid = t.schema()->id();
+    auto erm = t.get_effective_replication_map();
+    auto& tmap = erm->get_token_metadata_ptr()->tablets().get_tablet_map(tid);
+    const auto& topology = erm->get_topology();
+    std::unordered_map<locator::tablet_id, locator::tablet_info> maps;
+    co_await tmap.for_each_tablet([&] (locator::tablet_id id, const locator::tablet_info& info) {
+        maps.emplace(id, info);
+    });
```

Comment:
Looks like constructing maps is not necessary, you can build ranges directly here.

Reply:
I am using the map to avoid stalls. The func returns void. We could not yield.  For this case, we probably will not  have stalls in each func. But I've seen stalls in my other patches where we have a lot of work in each func. So I used the map anyway here.  

```
future<> tablet_map::for_each_tablet(seastar::noncopyable_function<void(tablet_id, const tablet_info&)> func) const {
    std::optional<tablet_id> tid = first_tablet();
    for (const tablet_info& ti : tablets()) {
        co_await coroutine::maybe_yield();
        func(*tid, ti);
        tid = next_tablet(*tid);
    }
}   
```

Reply:
> But I've seen stalls in my other patches where we have a lot of work in each func.

probably we should yield *in* the function, instead of in between them ?

and i concur with Tomek on this. the function returns `void`, but  `for_each_tablet()` does `co_await coroutine::maybe_yield();` as you quoted in the comment above. so does this imply that we can construct the `range` right in the function?

Reply:
I meant the case where a single call to func itself stalls. We need a version like for_each_tablet_gently where func returns a future and func can yield if needed.

Reply:
but the way how you process the `ranges` does not yield *in* the iteration, it yields at end of each iteration. this is identical to what we get from inlining the processing in `for_each_tablet()`.

Reply:
Yes, but why does it matter? The number of replicas should be low.  As I mentioned above: " For this case, we probably will not have stalls in each func.", even without any yield in the func. 

But I will add a for_each_tablet_gently.

Reply:
in the sense of correctness, it does not matter. but the following two functions might help to explain the difference:
```py
def sum_double_range(range):
  another_range = [e for e in range]
  return sum(2 * e for e in another_range)
```
and
```py
def sum_double_range(range):
  return sum(2 * e for e in range)
```

Reply:
The suggestion was not to add for_each_tablet_gently(), the existing for_each_tablet() is gentle enough, just put the computation in it to avoid intermediate map.